### PR TITLE
[16.0][FIX] account_statement_import_online_gocardless: IBAN comparison

### PR DESCRIPTION
**Pull request by ljsalvatierra-factorlibre**
It is possible that, when making the request to the `requisitions` endpoint, the IBAN bank account comes with a lower alphanumeric string. When comparing with the sanitized bank account (stored with upper) fails.

`self.journal_id.bank_account_id.sanitized_acc_number == account_data["iban"]`

to

`self.journal_id.bank_account_id.sanitized_acc_number == account_data["iban"].upper()`

I did refactor the method `_gocardless_finish_requisition` to be able to mock the requests made inside and create a unit test.

### CODE REVIEW

**Review by aritzolea — APPROVED** [review]

**Comment by ljsalvatierra-factorlibre** [comment]
Hi @pedrobaeza could you take a look please? I've seen that you have recently contributed to the same addon.

**Comment by pedrobaeza** [comment]
Please don't put `[WIP]` in the title, but use Draft status, as now, although you remove it, on mails, it still puts the WIP subject, so this makes people think that is not ready yet:

imagen


Can you please expose the problem this is handling? It should be always put on the initial comment, and also on the commit message.

**Review by MayteGLC — APPROVED** [review]

**Comment by ljsalvatierra-factorlibre** [comment]
> Please don't put `[WIP]` in the title, but use Draft status, as now, although you remove it, on mails, it still puts the WIP subject, so this makes people think that is not ready yet:
> 
> imagen
> 
> Can you please expose the problem this is handling? It should be always put on the initial comment, and also on the commit message.

I am aware, won't happen again, thank you.
I've updated the description, sorry for the confussion.

**Review by MayteGLC — APPROVED** [review]

**Comment by ljsalvatierra-factorlibre** [comment]
Hi @pedrobaeza is it ok to merge?

File: account_statement_import_online_gocardless/models/online_bank_statement_provider.py [hunk 1]
```diff
@@ -190,6 +190,39 @@ def action_check_gocardless_agreement(self):
...
+    def _gocardless_request_requisition(self):
+        response = requests.get(
+            f"{GOCARDLESS_ENDPOINT}/requisitions/{self.gocardless_requisition_id}/",
+            headers=self._gocardless_get_headers(),
+            timeout=REQUESTS_TIMEOUT,
+        )
+        if response.status_code == 200:
+            return json.loads(response.text)
+        else:
+            return {}
```
Comment (pedrobaeza) [inline account_statement_import_online_gocardless/models/online_bank_statement_provider.py]: And the same for the rest.

```suggestion
        if response.status_code == 200:
            return json.loads(response.text)
        return {}
```

**Review by pedrobaeza — COMMENTED** [review]
And please put the same explanation of the main comment in the commit expanded message (third line after a blank line and following)

File: account_statement_import_online_gocardless/models/online_bank_statement_provider.py (hunk 1 above)
Comment (ljsalvatierra-factorlibre) [reply account_statement_import_online_gocardless/models/online_bank_statement_provider.py]: Done :) thank you for your time!

**Review by pedrobaeza — COMMENTED** [review]
Sorry for asking for more modifications, but thinking twice, I think we can refactor this better, having only one method:

```python
    def _gocardless_request(self, path):
        response = requests.get(
            f"{GOCARDLESS_ENDPOINT}/path",
            headers=self._gocardless_get_headers(),
            timeout=REQUESTS_TIMEOUT,
        )
        if response.status_code == 200:
            return json.loads(response.text)
        return {}
```

and calling it with the multiple variants:

```python
self._gorcardless_request(f"requisitions/{self.gocardless_requisition_id}/")
self._gorcardless_request(f"accounts/{account_id}/")
...
```

My only doubt is the mock for working properly.

**Comment by ljsalvatierra-factorlibre** [comment]
> Sorry for asking for more modifications, but thinking twice, I think we can refactor this better, having only one method:
> 
> ```python
>     def _gocardless_request(self, path):
>         response = requests.get(
>             f"{GOCARDLESS_ENDPOINT}/path",
>             headers=self._gocardless_get_headers(),
>             timeout=REQUESTS_TIMEOUT,
>         )
>         if response.status_code == 200:
>             return json.loads(response.text)
>         return {}
> ```
> 
> and calling it with the multiple variants:
> 
> ```python
> self._gorcardless_request(f"requisitions/{self.gocardless_requisition_id}/")
> self._gorcardless_request(f"accounts/{account_id}/")
> ...
> ```
> 
> My only doubt is the mock for working properly.

I'll take a look thank you, I think it would be possible rewriting the `side_effect` method returning a different value depending on the arguments.

**Comment by pedrobaeza** [comment]
Any news on this?

**Comment by ljsalvatierra-factorlibre** [comment]
> Any news on this?

Hi, It's in my queue, but currently drowned on more urgent work, I'll try to take a look this week, sorry for the inconvenience.

**Comment by ljsalvatierra-factorlibre** [comment]
> Any news on this?

Hi @pedrobaeza sorry for the delay, please take a look and tell me what you think.

File: account_statement_import_online_gocardless/models/online_bank_statement_provider.py [hunk 2]
```diff
@@ -46,6 +46,44 @@ def _get_available_services(self):
...
+        response = requests.post(
+            url,
+            data=data,
+            headers=self._gocardless_get_headers(basic=True),
+            timeout=REQUESTS_TIMEOUT,
+        )
+
+        if response.status_code in [200, 201]:
+            content = json.loads(response.text)
+        return response, content
```
Comment (pedrobaeza) [inline account_statement_import_online_gocardless/models/online_bank_statement_provider.py]: Simplify it in one method:

```suggestion
    def _gocardless_request(self, request_type="get", endpoint, params=None):
        content = {}
        url = url_join(GOCARDLESS_API, endpoint)
        response = getattr(request, request_type)(
            url,
            params=params,
            headers=self._gocardless_get_headers(),
            timeout=REQUESTS_TIMEOUT,
        )
        if response.status_code == 200:
            content = json.loads(response.text)
        return response, content
```

And add the extra argument of each of the other calls

**Comment by pedrobaeza** [comment]
Please squash comments into one. It's better to amend the previous comment than adding one extra. And finally, `[UPD]` is not a valid commit tag.

**Comment by ljsalvatierra-factorlibre** [comment]
> Please squash comments into one. It's better to amend the previous comment than adding one extra. And finally, `[UPD]` is not a valid commit tag.

Done

**Review by pedrobaeza — APPROVED** [review]
Thanks for the refactoring:

/ocabot merge minor

**Pull request merged**